{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "914fc79d",
   "metadata": {},
   "source": [
    "# Replogle dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6465cff",
   "metadata": {},
   "source": [
    "Descriptive statistics and Preprocessing of replogle dataset to be used as a perturbation datasets for training bmfm-rna models.   \n",
    "We use the original file with raw counts, but apply the split from scGPT.  \n",
    "A small file with few perturbations but in same format is created for testing.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e9b23c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e30c944",
   "metadata": {},
   "source": [
    "We are using the scGPT processed version and splits of Repologle dataset from [scGPT paper](https://www.nature.com/articles/s41592-024-02201-0) and supplementary.\n",
    "```\n",
    "Replogle. The Replogle perturbation dataset contains genome-wide\n",
    "perturbations in the K562 leukemia cell line with CRISPR interference34.\n",
    "With consideration for data quality, we retained the subset of data\n",
    "matching the 1,973 perturbations identified in the original study with\n",
    "strong transcriptional phenotypes. We also removed 150 perturbations\n",
    "that had no expression records of the perturbed gene in the sequencing data. We further retained 100 samples per perturbation and 2,500\n",
    "control samples. The processed whole dataset consists of 171,542 samples from 1,823 perturbations, among which 99 are perturbations of\n",
    "transcription factors. The test set consists of 456 perturbations, among\n",
    "which 25 are perturbations of transcription factors.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "25eddab7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/dccstor/bmfm-targets1/users/sivanra/miniforge/miniforge/envs/targets072025/lib/python3.11/site-packages/louvain/__init__.py:54: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import get_distribution, DistributionNotFound\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "85750126",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dataset_folder = \"/proj/bmfm/datasets/vcc/replogle2022/scGPT_Fig3_AB_PerturbPred/k562_1900_100_re_ctrl_sample\"\n",
    "dataset_folder = \"/dccstor/bmfm-targets/data/omics/transcriptome/scRNA/finetune/Perturbation/replogle/k562_1900_100_re_ctrl_sample\"\n",
    "filename = \"perturb_processed.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c434e69a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 171542 × 6017\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'batch', 'condition', 'cell_type', 'dose_val', 'control', 'condition_name'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'log1p', 'non_dropout_gene_idx', 'non_zeros_gene_idx', 'rank_genes_groups_cov_all', 'top_non_dropout_de_20', 'top_non_zero_de_20'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(Path(dataset_folder) / filename)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dcbf1c49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 387347629 stored elements and shape (171542, 6017)>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca74751a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript',\n",
       "       'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI',\n",
       "       'core_scale_factor', 'core_adjusted_UMI_count', 'batch', 'condition',\n",
       "       'cell_type', 'dose_val', 'control', 'condition_name'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cb5cf014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "NFRKB             100\n",
       "NCBP2             100\n",
       "NBAS              100\n",
       "NARS2             100\n",
       "                 ... \n",
       "RPS27A             26\n",
       "PDCD11             26\n",
       "HAUS1              26\n",
       "PRPF4              25\n",
       "SEM1               21\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['gene'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf9a8de4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "control\n",
       "0    169042\n",
       "1      2500\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['control'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3cff5506",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "RANBP1              0\n",
       "RAMAC               0\n",
       "RAI14               0\n",
       "RAE1                0\n",
       "                 ... \n",
       "HIRA                0\n",
       "HIPK3               0\n",
       "HIPK1               0\n",
       "HINFP               0\n",
       "HNRNPK              0\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs['control'] ==1]['gene'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c656a1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "batch\n",
       "1969    2500\n",
       "1037     100\n",
       "1005     100\n",
       "1003     100\n",
       "1002     100\n",
       "        ... \n",
       "1461      26\n",
       "1138      26\n",
       "626       26\n",
       "1252      25\n",
       "1510      21\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['batch'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb327c6b",
   "metadata": {},
   "source": [
    "### Apply split from scGPT file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e7cb41f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "splits_folder = \"splits\"\n",
    "splits_file = \"k562_1900_100_re_ctrl_sample_simulation_1_0.75.pkl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bf5aafee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test count: 456\n",
      "['ACTB+ctrl', 'ADAM10+ctrl', 'ADNP+ctrl', 'ADRM1+ctrl', 'ADSL+ctrl', 'ALDOA+ctrl', 'ANKRD27+ctrl', 'ANKRD39+ctrl', 'ARHGAP6+ctrl', 'ASCC3+ctrl']\n",
      "train count: 1230\n",
      "['AACS+ctrl', 'AAMP+ctrl', 'AAR2+ctrl', 'AARS+ctrl', 'AARS2+ctrl', 'AATF+ctrl', 'ABCB7+ctrl', 'ABCE1+ctrl', 'ABCF1+ctrl', 'ABT1+ctrl']\n",
      "val count: 137\n",
      "['ACTL6A+ctrl', 'ADO+ctrl', 'AHCTF1+ctrl', 'ATR+ctrl', 'BTAF1+ctrl', 'CCNL1+ctrl', 'CCNQ+ctrl', 'CDIPT+ctrl', 'CDK9+ctrl', 'CENPC+ctrl']\n"
     ]
    }
   ],
   "source": [
    "with open(Path(dataset_folder) / splits_folder / splits_file, \"rb\") as f:\n",
    "    splits = pickle.load(f)\n",
    "    for split_name, gene_list in splits.items():\n",
    "        print(f\"{split_name} count: {len(gene_list)}\")\n",
    "        print(gene_list[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7064f1cf",
   "metadata": {},
   "source": [
    "Lets create a column in adata with the split:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "308e8f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_split = {}\n",
    "for split_name, gene_list in splits.items():\n",
    "    gene_to_split.update({gene.replace(\"+ctrl\",\"\"): split_name for gene in gene_list})\n",
    "\n",
    "# Map the \"gene\" column in adata.obs to split labels\n",
    "adata.obs[\"scgpt_split\"] = adata.obs[\"gene\"].map(gene_to_split)\n",
    "\n",
    "#rename val to dev\n",
    "adata.obs.loc[adata.obs[\"scgpt_split\"] == \"val\", \"scgpt_split\"] = \"dev\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "528dadf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scgpt_split\n",
       "train    113743\n",
       "test      42351\n",
       "dev       12948\n",
       "NaN        2500\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"scgpt_split\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fe41bcc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "ZNHIT6    100\n",
       "AACS      100\n",
       "AAMP      100\n",
       "ZNF787    100\n",
       "ZNF706    100\n",
       "         ... \n",
       "SARS2       0\n",
       "SBNO1       0\n",
       "ESF1        0\n",
       "ETF1        0\n",
       "POLR3C      0\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs[\"scgpt_split\"]==\"train\"][\"gene\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "db0dae31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(62.393307734503566)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs[\"scgpt_split\"]==\"train\"][\"gene\"].value_counts().mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fe6cc2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to save again\n",
    "\n",
    "#adata.write(\"/dccstor/bmfm-targets/data/omics/transcriptome/scRNA/finetune/Perturbation/replogle/scGPT_k562_1900_100_re_ctrl_sample_perturb_processed_with_split.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "636b849f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>class</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>in_matrix</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>cv</th>\n",
       "      <th>fano</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000237491</th>\n",
       "      <td>LINC01409</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778747</td>\n",
       "      <td>810065</td>\n",
       "      <td>gene_version10</td>\n",
       "      <td>+</td>\n",
       "      <td>31318</td>\n",
       "      <td>True</td>\n",
       "      <td>0.116626</td>\n",
       "      <td>0.349971</td>\n",
       "      <td>3.000803</td>\n",
       "      <td>1.050194</td>\n",
       "      <td>True</td>\n",
       "      <td>0.113362</td>\n",
       "      <td>0.325672</td>\n",
       "      <td>-0.055866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <td>KLHL17</td>\n",
       "      <td>chr1</td>\n",
       "      <td>960584</td>\n",
       "      <td>965719</td>\n",
       "      <td>gene_version14</td>\n",
       "      <td>+</td>\n",
       "      <td>5135</td>\n",
       "      <td>True</td>\n",
       "      <td>0.105599</td>\n",
       "      <td>0.330678</td>\n",
       "      <td>3.131439</td>\n",
       "      <td>1.035497</td>\n",
       "      <td>False</td>\n",
       "      <td>0.092125</td>\n",
       "      <td>0.245210</td>\n",
       "      <td>-0.637191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>HES4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>998962</td>\n",
       "      <td>1000172</td>\n",
       "      <td>gene_version10</td>\n",
       "      <td>-</td>\n",
       "      <td>1210</td>\n",
       "      <td>True</td>\n",
       "      <td>0.242700</td>\n",
       "      <td>0.550596</td>\n",
       "      <td>2.268630</td>\n",
       "      <td>1.249098</td>\n",
       "      <td>True</td>\n",
       "      <td>0.209183</td>\n",
       "      <td>0.462903</td>\n",
       "      <td>0.935613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <td>ISG15</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1001138</td>\n",
       "      <td>1014540</td>\n",
       "      <td>gene_version10</td>\n",
       "      <td>+</td>\n",
       "      <td>13402</td>\n",
       "      <td>True</td>\n",
       "      <td>0.334181</td>\n",
       "      <td>0.675689</td>\n",
       "      <td>2.021921</td>\n",
       "      <td>1.366189</td>\n",
       "      <td>True</td>\n",
       "      <td>0.307255</td>\n",
       "      <td>0.716577</td>\n",
       "      <td>2.768372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000078808</th>\n",
       "      <td>SDF4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1216908</td>\n",
       "      <td>1232067</td>\n",
       "      <td>gene_version17</td>\n",
       "      <td>-</td>\n",
       "      <td>15159</td>\n",
       "      <td>True</td>\n",
       "      <td>0.889459</td>\n",
       "      <td>1.054260</td>\n",
       "      <td>1.185283</td>\n",
       "      <td>1.249596</td>\n",
       "      <td>True</td>\n",
       "      <td>0.628492</td>\n",
       "      <td>0.370973</td>\n",
       "      <td>0.093641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>chrM</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>gene_version2</td>\n",
       "      <td>+</td>\n",
       "      <td>1811</td>\n",
       "      <td>True</td>\n",
       "      <td>24.573463</td>\n",
       "      <td>12.764438</td>\n",
       "      <td>0.519440</td>\n",
       "      <td>6.630358</td>\n",
       "      <td>True</td>\n",
       "      <td>3.177331</td>\n",
       "      <td>1.517886</td>\n",
       "      <td>1.775934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>gene_version2</td>\n",
       "      <td>-</td>\n",
       "      <td>524</td>\n",
       "      <td>True</td>\n",
       "      <td>6.160370</td>\n",
       "      <td>5.194642</td>\n",
       "      <td>0.843235</td>\n",
       "      <td>4.380306</td>\n",
       "      <td>True</td>\n",
       "      <td>1.858941</td>\n",
       "      <td>1.337521</td>\n",
       "      <td>2.124659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278704</th>\n",
       "      <td>BX004987.1</td>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>56140</td>\n",
       "      <td>58376</td>\n",
       "      <td>gene_version1</td>\n",
       "      <td>-</td>\n",
       "      <td>2236</td>\n",
       "      <td>True</td>\n",
       "      <td>0.195008</td>\n",
       "      <td>0.455099</td>\n",
       "      <td>2.333745</td>\n",
       "      <td>1.062085</td>\n",
       "      <td>True</td>\n",
       "      <td>0.183181</td>\n",
       "      <td>0.353474</td>\n",
       "      <td>0.144999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278384</th>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>GL000218.1</td>\n",
       "      <td>51867</td>\n",
       "      <td>54893</td>\n",
       "      <td>gene_version1</td>\n",
       "      <td>-</td>\n",
       "      <td>3026</td>\n",
       "      <td>True</td>\n",
       "      <td>0.190869</td>\n",
       "      <td>0.450850</td>\n",
       "      <td>2.362089</td>\n",
       "      <td>1.064948</td>\n",
       "      <td>True</td>\n",
       "      <td>0.184353</td>\n",
       "      <td>0.397784</td>\n",
       "      <td>0.465131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>4612</td>\n",
       "      <td>29626</td>\n",
       "      <td>gene_version6</td>\n",
       "      <td>-</td>\n",
       "      <td>25014</td>\n",
       "      <td>True</td>\n",
       "      <td>0.163767</td>\n",
       "      <td>0.427101</td>\n",
       "      <td>2.607982</td>\n",
       "      <td>1.113872</td>\n",
       "      <td>True</td>\n",
       "      <td>0.144298</td>\n",
       "      <td>0.352038</td>\n",
       "      <td>0.134626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6017 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_name         chr    start      end           class  \\\n",
       "gene_id                                                                     \n",
       "ENSG00000237491   LINC01409        chr1   778747   810065  gene_version10   \n",
       "ENSG00000187961      KLHL17        chr1   960584   965719  gene_version14   \n",
       "ENSG00000188290        HES4        chr1   998962  1000172  gene_version10   \n",
       "ENSG00000187608       ISG15        chr1  1001138  1014540  gene_version10   \n",
       "ENSG00000078808        SDF4        chr1  1216908  1232067  gene_version17   \n",
       "...                     ...         ...      ...      ...             ...   \n",
       "ENSG00000198786      MT-ND5        chrM    12337    14148   gene_version2   \n",
       "ENSG00000198695      MT-ND6        chrM    14149    14673   gene_version2   \n",
       "ENSG00000278704  BX004987.1  GL000009.2    56140    58376   gene_version1   \n",
       "ENSG00000278384  AL354822.1  GL000218.1    51867    54893   gene_version1   \n",
       "ENSG00000271254  AC240274.1  KI270711.1     4612    29626   gene_version6   \n",
       "\n",
       "                strand  length  in_matrix       mean        std        cv  \\\n",
       "gene_id                                                                     \n",
       "ENSG00000237491      +   31318       True   0.116626   0.349971  3.000803   \n",
       "ENSG00000187961      +    5135       True   0.105599   0.330678  3.131439   \n",
       "ENSG00000188290      -    1210       True   0.242700   0.550596  2.268630   \n",
       "ENSG00000187608      +   13402       True   0.334181   0.675689  2.021921   \n",
       "ENSG00000078808      -   15159       True   0.889459   1.054260  1.185283   \n",
       "...                ...     ...        ...        ...        ...       ...   \n",
       "ENSG00000198786      +    1811       True  24.573463  12.764438  0.519440   \n",
       "ENSG00000198695      -     524       True   6.160370   5.194642  0.843235   \n",
       "ENSG00000278704      -    2236       True   0.195008   0.455099  2.333745   \n",
       "ENSG00000278384      -    3026       True   0.190869   0.450850  2.362089   \n",
       "ENSG00000271254      -   25014       True   0.163767   0.427101  2.607982   \n",
       "\n",
       "                     fano  highly_variable     means  dispersions  \\\n",
       "gene_id                                                             \n",
       "ENSG00000237491  1.050194             True  0.113362     0.325672   \n",
       "ENSG00000187961  1.035497            False  0.092125     0.245210   \n",
       "ENSG00000188290  1.249098             True  0.209183     0.462903   \n",
       "ENSG00000187608  1.366189             True  0.307255     0.716577   \n",
       "ENSG00000078808  1.249596             True  0.628492     0.370973   \n",
       "...                   ...              ...       ...          ...   \n",
       "ENSG00000198786  6.630358             True  3.177331     1.517886   \n",
       "ENSG00000198695  4.380306             True  1.858941     1.337521   \n",
       "ENSG00000278704  1.062085             True  0.183181     0.353474   \n",
       "ENSG00000278384  1.064948             True  0.184353     0.397784   \n",
       "ENSG00000271254  1.113872             True  0.144298     0.352038   \n",
       "\n",
       "                 dispersions_norm  \n",
       "gene_id                            \n",
       "ENSG00000237491         -0.055866  \n",
       "ENSG00000187961         -0.637191  \n",
       "ENSG00000188290          0.935613  \n",
       "ENSG00000187608          2.768372  \n",
       "ENSG00000078808          0.093641  \n",
       "...                           ...  \n",
       "ENSG00000198786          1.775934  \n",
       "ENSG00000198695          2.124659  \n",
       "ENSG00000278704          0.144999  \n",
       "ENSG00000278384          0.465131  \n",
       "ENSG00000271254          0.134626  \n",
       "\n",
       "[6017 rows x 16 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7d0bcf0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 1.0932776 , 0.        , 1.0932776 ],\n",
       "       [0.        , 0.        , 0.        , 0.6340871 , 0.6340871 ,\n",
       "        0.        , 1.2963425 , 0.        , 0.        , 1.5131915 ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 1.5427817 ,\n",
       "        0.        , 0.5513052 , 0.        , 0.9046365 , 0.9046365 ],\n",
       "       [0.        , 0.        , 0.5408124 , 0.8898659 , 0.        ,\n",
       "        0.        , 0.5408124 , 0.        , 0.5408124 , 0.8898659 ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 1.178307  ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.8864786 , 0.53841245, 0.        , 0.53841245],\n",
       "       [0.        , 0.        , 0.        , 0.75866437, 1.1850482 ,\n",
       "        0.        , 0.        , 0.75866437, 0.        , 0.75866437],\n",
       "       [0.        , 0.        , 0.        , 0.7561039 , 0.7561039 ,\n",
       "        0.        , 0.        , 0.        , 0.7561039 , 1.3415332 ],\n",
       "       [0.        , 0.        , 0.6179556 , 0.        , 0.99704576,\n",
       "        0.        , 0.        , 0.        , 0.99704576, 0.6179556 ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.96512437,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ]],\n",
       "      dtype=float32)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.toarray()[:10,:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d60a27ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 171542 × 6017\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'batch', 'condition', 'cell_type', 'dose_val', 'control', 'condition_name', 'scgpt_split'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'log1p', 'non_dropout_gene_idx', 'non_zeros_gene_idx', 'rank_genes_groups_cov_all', 'top_non_dropout_de_20', 'top_non_zero_de_20'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8e2d87f",
   "metadata": {},
   "source": [
    "### Save a small file for tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1dff0c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "limit_genes=3\n",
    "split_short_list = {}\n",
    "for split_name, gene_list in splits.items():\n",
    "    split_short_list.update({gene.replace(\"+ctrl\",\"\"): split_name for gene in gene_list[:limit_genes]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "61f25b1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACTB': 'test',\n",
       " 'ADAM10': 'test',\n",
       " 'ADNP': 'test',\n",
       " 'AACS': 'train',\n",
       " 'AAMP': 'train',\n",
       " 'AAR2': 'train',\n",
       " 'ACTL6A': 'val',\n",
       " 'ADO': 'val',\n",
       " 'AHCTF1': 'val'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_short_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d3de1817",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_small = adata[adata.obs[\"gene\"].isin(set(split_short_list.keys()|{\"non-targeting\"}))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "66cf7bae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 3339 × 6017\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'batch', 'condition', 'cell_type', 'dose_val', 'control', 'condition_name', 'scgpt_split'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'log1p', 'non_dropout_gene_idx', 'non_zeros_gene_idx', 'rank_genes_groups_cov_all', 'top_non_dropout_de_20', 'top_non_zero_de_20'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "903c57b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scgpt_split\n",
       "train    293\n",
       "test     284\n",
       "dev      262\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small.obs[\"scgpt_split\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "57059f6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "AACS              100\n",
       "AAMP              100\n",
       "ACTB              100\n",
       "ADNP              100\n",
       "ADO               100\n",
       "AHCTF1            100\n",
       "AAR2               93\n",
       "ADAM10             84\n",
       "ACTL6A             62\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small.obs[\"gene\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "24a1e75e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to save again\n",
    "#adata_small.write(\"/dccstor/bmfm-targets/data/omics/transcriptome/scRNA/finetune/Perturbation/replogle/scGPT_k562_1900_100_re_ctrl_sample_perturb_processed_with_split_small.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59d74d03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 7650274 stored elements and shape (3339, 6017)>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47c69361",
   "metadata": {},
   "source": [
    "## Create row counts file with same split\n",
    "After experimenting we decided to use raw counts. \n",
    "Merge with the original raw h5ad to get raw counts instead of scGPT normalized, but keep the split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2d9a6e50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CATCCCACACAGTGAG-99-0', 'TCCACCACAGATGCGA-127-0',\n",
       "       'CCTTTGGTCTGTCGTC-250-0', 'CATCCCAGTGAGCGAT-24-0',\n",
       "       'GTCTACCGTAGCACAG-186-0', 'GCCATGGTCCGTGGCA-88-0',\n",
       "       'ACAGGGATCACGAACT-101-0', 'TGCACGGCAAAGGGCT-181-0',\n",
       "       'GTTACCCTCGTGTCAA-104-0', 'GAAGCGATCCTTTGAT-148-0',\n",
       "       ...\n",
       "       'GGGCGTTAGATGGTCG-98-1969', 'CTTCTAACAGAGTCAG-119-1969',\n",
       "       'CACGTTCCACGGTGTC-103-1969', 'GTCAGCGTCCTTGGAA-149-1969',\n",
       "       'GAGCTGCCACGACTAT-100-1969', 'TGTCCTGCACACCGCA-168-1969',\n",
       "       'TCCCAGTCAATAGTCC-56-1969', 'CTACCTGCATAACGGG-75-1969',\n",
       "       'TCCGGGAAGTAGCCAG-96-1969', 'GCCAGCACACTAGGCC-84-1969'],\n",
       "      dtype='object', name='cell_barcode', length=171542)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs_names "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ebfed542",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_counts_filename = \"K562_gwps_raw_singlecell_01.h5ad\"\n",
    "raw_counts_subset_filename = \"K562_gwps_raw_singlecell_01.subset.h5ad\"\n",
    "\n",
    "adata_raw_counts = sc.read_h5ad(Path(dataset_folder).parent /\"K562_gwps_raw_singlecell_01.h5ad\", backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "30bc9413",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAACCCAAGAAACCAT-157', 'AAACCCAAGAAACCAT-207', 'AAACCCAAGAAACCAT-29',\n",
       "       'AAACCCAAGAAAGCGA-149', 'AAACCCAAGAAATCCA-172', 'AAACCCAAGAACAAGG-215',\n",
       "       'AAACCCAAGAACAGGA-209', 'AAACCCAAGAACAGGA-54', 'AAACCCAAGAACCCGA-228',\n",
       "       'AAACCCAAGAACCCGA-48',\n",
       "       ...\n",
       "       'TTTGTTGTCTTGGCTC-166', 'TTTGTTGTCTTGGCTC-86', 'TTTGTTGTCTTGTGCC-179',\n",
       "       'TTTGTTGTCTTGTTAC-124', 'TTTGTTGTCTTGTTAC-246', 'TTTGTTGTCTTTCCAA-201',\n",
       "       'TTTGTTGTCTTTCGAT-79', 'TTTGTTGTCTTTCTAG-218', 'TTTGTTGTCTTTGCGC-169',\n",
       "       'TTTGTTGTCTTTGCGC-213'],\n",
       "      dtype='object', name='cell_barcode', length=1989578)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw_counts.obs_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c36266fa",
   "metadata": {},
   "source": [
    "looks like we need to trim adata.obs_names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9b0b0ce3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CATCCCACACAGTGAG-99',\n",
       " 'TCCACCACAGATGCGA-127',\n",
       " 'CCTTTGGTCTGTCGTC-250',\n",
       " 'CATCCCAGTGAGCGAT-24',\n",
       " 'GTCTACCGTAGCACAG-186',\n",
       " 'GCCATGGTCCGTGGCA-88',\n",
       " 'ACAGGGATCACGAACT-101',\n",
       " 'TGCACGGCAAAGGGCT-181',\n",
       " 'GTTACCCTCGTGTCAA-104',\n",
       " 'GAAGCGATCCTTTGAT-148']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_names = [s.rsplit(\"-\", 1)[0] for s in adata.obs_names]\n",
    "cell_names[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "20a9c188",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAACCCAAGACAACAT-110',\n",
       " 'AAACCCAAGACATCCT-144',\n",
       " 'AAACCCAAGAGCATAT-255',\n",
       " 'AAACCCAAGAGCCGAT-22',\n",
       " 'AAACCCAAGAGGTCAC-12']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check on first 100 cell names if a subset exists:\n",
    "[cell for cell in adata_raw_counts.obs_names[:100] if cell in cell_names]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "afad5060",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "171542"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_names_set = set(cell_names)  \n",
    "mask = adata_raw_counts.obs_names.isin(cell_names_set) \n",
    "keep_indices = mask.nonzero()[0]  \n",
    "len(keep_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "319de081",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "171542"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(cell_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "226d65fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#workes but takes 18 min\n",
    "#adata_raw_counts[keep_indices, :].copy(Path(dataset_folder).parent / raw_counts_subset_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fedeb1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to save again\n",
    "# adata_raw_counts_subset.write_h5ad(Path(dataset_folder).parent / raw_counts_subset_filename, compression=\"gzip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6bb37f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_h5ad = sc.read_h5ad(Path(dataset_folder).parent / raw_counts_subset_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "468e61e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 171542 × 8248\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "72a7a32e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript',\n",
       "       'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI',\n",
       "       'core_scale_factor', 'core_adjusted_UMI_count'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_h5ad.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2bff49bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "NFRKB             100\n",
       "NCBP2             100\n",
       "NBAS              100\n",
       "NARS2             100\n",
       "                 ... \n",
       "RPS27A             26\n",
       "PDCD11             26\n",
       "HAUS1              26\n",
       "PRPF4              25\n",
       "SEM1               21\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_h5ad.obs[\"gene\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8e14c5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy split again\n",
    "# Map the \"gene\" column in adata.obs to split labels\n",
    "subset_h5ad.obs[\"scgpt_split\"] = subset_h5ad.obs[\"gene\"].map(gene_to_split)\n",
    "\n",
    "#rename val to dev\n",
    "subset_h5ad.obs.loc[subset_h5ad.obs[\"scgpt_split\"] == \"val\", \"scgpt_split\"] = \"dev\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6e46b8f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scgpt_split\n",
       "train    113743\n",
       "test      42351\n",
       "dev       12948\n",
       "NaN        2500\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_h5ad.obs[\"scgpt_split\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "90c2ad2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to save again\n",
    "# save final h5ad\n",
    "#subset_h5ad.write_h5ad(Path(dataset_folder).parent / \"k562_gwps_raw_singlecell_01_processed_with_scgpt_split.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "74f75e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_h5ad_small = subset_h5ad[subset_h5ad.obs[\"gene\"].isin(set(split_short_list.keys()|{\"non-targeting\"}))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a03a28f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to save again\n",
    "#subset_h5ad_small.write_h5ad(Path(dataset_folder).parent / \"k562_gwps_raw_singlecell_01_processed_with_scgpt_split_small.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b39503f",
   "metadata": {},
   "source": [
    "#### Save gene names to csv file, this is usefull for external evalautions like cell-eval (https://pypi.org/project/cell-eval/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "44c3f641",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names_file = \"gene_names.csv\"\n",
    "genes = subset_h5ad.obs[\"gene\"].unique()\n",
    "gene_names = [gene for gene in genes if gene != 'non-targeting']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2718ae8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INO80B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MRPL52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNRPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FUNDC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1817</th>\n",
       "      <td>CPSF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1818</th>\n",
       "      <td>GTF2F2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1819</th>\n",
       "      <td>NFYC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1820</th>\n",
       "      <td>PDCD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1821</th>\n",
       "      <td>EMG1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1822 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene\n",
       "0     INO80B\n",
       "1     MRPL52\n",
       "2       LTV1\n",
       "3      SNRPC\n",
       "4     FUNDC2\n",
       "...      ...\n",
       "1817   CPSF2\n",
       "1818  GTF2F2\n",
       "1819    NFYC\n",
       "1820  PDCD11\n",
       "1821    EMG1\n",
       "\n",
       "[1822 rows x 1 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = pd.DataFrame(gene_names, columns=[\"gene\"])\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c82099",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to CSV\n",
    "genes_df.to_csv(Path(dataset_folder).parent / gene_names_file, index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1beab55c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save test set to a seaparate file to be used as groun-truth in comparisons\n",
    "gt_file_name = \"\"\n",
    "small_gt_file_name = \"\"\n",
    "gt_subset = subset_h5ad[subset_h5ad.obs[\"scgpt_split\"] == \"test\"]\n",
    "gt_subset.write_h5ad(filename=Path(dataset_folder).parent / gt_file_name)\n",
    "\n",
    "gt_subset_small = subset_h5ad_small[subset_h5ad_small.obs[\"scgpt_split\"] == \"test\"]\n",
    "gt_subset_small.write_h5ad(filename=Path(dataset_folder).parent / small_gt_file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7fe0fbc",
   "metadata": {},
   "source": [
    "# Validate final files\n",
    "Run this part only to read the final files created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "35866251",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "42e9a765",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_folder = \"/dccstor/bmfm-targets/data/omics/transcriptome/scRNA/finetune/Perturbation/replogle/k562_1900_100_re_ctrl_sample\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "bfe9fef9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3339 × 8248\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'scgpt_split'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small = sc.read_h5ad(Path(dataset_folder).parent / \"k562_gwps_raw_singlecell_01_processed_with_scgpt_split_small.h5ad\")\n",
    "adata_small"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "18328ce4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "AACS              100\n",
       "AAMP              100\n",
       "ACTB              100\n",
       "ADNP              100\n",
       "ADO               100\n",
       "AHCTF1            100\n",
       "AAR2               93\n",
       "ADAM10             84\n",
       "ACTL6A             62\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small.obs[\"gene\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "de6ab57e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scgpt_split\n",
       "NaN      2500\n",
       "train     293\n",
       "test      284\n",
       "dev       262\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_small.obs[\"scgpt_split\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d1a8697a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 171542 × 8248\n",
       "    obs: 'gem_group', 'gene', 'gene_id', 'transcript', 'gene_transcript', 'sgID_AB', 'mitopercent', 'UMI_count', 'z_gemgroup_UMI', 'core_scale_factor', 'core_adjusted_UMI_count', 'scgpt_split'\n",
       "    var: 'gene_name', 'chr', 'start', 'end', 'class', 'strand', 'length', 'in_matrix', 'mean', 'std', 'cv', 'fano'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(Path(dataset_folder).parent / \"k562_gwps_raw_singlecell_01_processed_with_scgpt_split.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "143a800a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "non-targeting    2500\n",
       "NFRKB             100\n",
       "NCBP2             100\n",
       "NBAS              100\n",
       "NARS2             100\n",
       "                 ... \n",
       "RPS27A             26\n",
       "PDCD11             26\n",
       "HAUS1              26\n",
       "PRPF4              25\n",
       "SEM1               21\n",
       "Name: count, Length: 1823, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"gene\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ba6eddad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scgpt_split\n",
       "train    113743\n",
       "test      42351\n",
       "dev       12948\n",
       "NaN        2500\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"scgpt_split\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "03917efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "counts = pd.crosstab(\n",
    "    adata.obs[\"gene\"].astype(str),\n",
    "    adata.obs[\"scgpt_split\"].astype(str),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "01eaa1d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\n",
      "AACS      100\n",
      "AAMP      100\n",
      "AAR2       93\n",
      "AARS       46\n",
      "AARS2     100\n",
      "         ... \n",
      "ZNF574    100\n",
      "ZNF695    100\n",
      "ZNF706    100\n",
      "ZNF787    100\n",
      "ZNHIT6    100\n",
      "Name: train, Length: 1229, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(counts[\"train\"][counts[\"train\"]>0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "8ff70761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\n",
      "ACTL6A     62\n",
      "ADO       100\n",
      "AHCTF1    100\n",
      "ATR        78\n",
      "BTAF1     100\n",
      "         ... \n",
      "WAC       100\n",
      "WDR26     100\n",
      "WDR46     100\n",
      "YBX3      100\n",
      "ZNF622     92\n",
      "Name: dev, Length: 137, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(counts[\"dev\"][counts[\"dev\"]>0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "735f415d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\n",
      "ACTB      100\n",
      "ADAM10     84\n",
      "ADNP      100\n",
      "ADRM1     100\n",
      "ADSL      100\n",
      "         ... \n",
      "ZNF84     100\n",
      "ZNHIT1    100\n",
      "ZNHIT3    100\n",
      "ZNRD1      81\n",
      "ZW10      100\n",
      "Name: test, Length: 456, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(counts[\"test\"][counts[\"test\"]>0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "targets072025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
